### PR TITLE
perf: Reduce memory footprint for ChunkedDistribuedLogProb

### PR DESCRIPTION
**Pull request by nujoug**
# What does this PR do ?

**Reduces the peak memory footprint when using chunk during loss function. (when sequence packing is disabled)**

# Issues
[Improve loss function memory usage ](https://github.com/NVIDIA-NeMo/RL/issues/990)

# Details
## Problem
Current approach stores each chunk's gradient in a list, followed by `torch.cat` at the end to return the entire gradient tensor. This resulted in at least 2 copies of gradient tensor to exist during peak.

Screenshot 2026-02-06 at 10 45 31 AM

## Modification
Preallocate a gradient tensor and copy each chunk's gradient inplace.

Screenshot 2026-02-06 at 10 57 57 AM

## Additional Gains
Reducing the chunk size observes less reduction in peak memory in current approach due to some intermediate tensors overlapping (lazy delete).

Screenshot 2026-02-06 at 11 09 39 AM
Screenshot 2026-02-06 at 11 14 37 AM

---
With explicit deallocation of intermediate tensor (using `del`), the memory footprint reduction is more significant (`-0.4GiB` to `-0.6GiB`)

Screenshot 2026-02-06 at 11 28 39 AM
Screenshot 2026-02-06 at 11 27 07 AM

## Caveats
The modification does not reduce the peak memory when sequence packing is enable. This is because [SequencePackingLossWrapper](https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/algorithms/loss_functions.py#L870) is being used and default `torch.autograd` handles the backprop resulting in this undesirable behavior.

Screenshot 2026-02-06 at 11 36 53 AM

This [issue](https://github.com/NVIDIA-NeMo/RL/issues/1247) should be able to solve this once there is a customize `torch.autograd.Function` that can handle sequence packing.

## Summary by CodeRabbit

* **Performance Improvements**
  * Reduced memory use and improved efficiency during backward/gradient computation by switching to preallocated in-place updates and explicit cleanup, lowering peak memory and improving stability.

* **Tests**
  * Added comprehensive distributed/chunked entropy tests (forward and optional backward) with multi-GPU validation to ensure numerical parity with baseline.

### CODE REVIEW

File: nemo_rl/distributed/model_utils.py [hunk 1]
```diff
@@ -220,7 +220,7 @@ def backward(
         seq_size = int(vocab_parallel_logits.shape[1])
         num_chunks = (seq_size + chunk_size - 1) // chunk_size
 
-        all_grad_input = []
+        grad_input: torch.Tensor = torch.empty_like(vocab_parallel_logits)
```
[the remark this answers is not in the document]
Comment (nujoug) [reply nemo_rl/distributed/model_utils.py]: good catch

**Comment by nujoug** [comment]
@terrykong 

Hi Terry, can you look at the codecov/patch failure? `tests/unit/distributed/test_model_utils.py` should have already covered the highlighted lines. 

Thanks

File: nemo_rl/distributed/model_utils.py [hunk 2]
```diff
@@ -220,7 +220,9 @@ def backward(
         seq_size = int(vocab_parallel_logits.shape[1])
         num_chunks = (seq_size + chunk_size - 1) // chunk_size
 
-        all_grad_input = []
+        grad_input: torch.Tensor = torch.empty_like(
```
Comment (guyueh1) [inline nemo_rl/distributed/model_utils.py]:
```suggestion
        grad_input: torch.Tensor = torch.zeros_like(
```

File: nemo_rl/distributed/model_utils.py [hunk 3]
```diff
@@ -326,7 +333,10 @@ def backward(
 
         B, S, V_local = vocab_parallel_logits.shape
         num_chunks = (int(S) + chunk_size - 1) // chunk_size
-        all_grad_input: list[torch.Tensor] = []
+
+        grad_input: torch.Tensor = torch.empty_like(
```
Comment (guyueh1) [inline nemo_rl/distributed/model_utils.py]:
```suggestion
        grad_input: torch.Tensor = torch.zeros_like(
```
